iroha/src/sumeragi/view_change.rs

```diff
@@ -0,0 +1,325 @@
+//! Structures related to proofs and reasons of view changes.
+//! Where view change is a process of changing topology due to some faulty network behavior.
+
+use std::{collections::HashSet, fmt::Display};
+
+use iroha_crypto::{Hash, KeyPair, PublicKey, Signature, Signatures};
+use iroha_data_model::prelude::PeerId;
+use iroha_derive::*;
+use iroha_error::Result;
+use parity_scale_codec::{Decode, Encode};
+
+use super::message::TransactionReceipt;
+use crate::block::EmptyChainHash;
+
+/// The proof of a view change. It needs to be signed by f+1 peers for proof to be valid and view change to happen.
+#[derive(Clone, Debug, Io, Encode, Decode)]
+pub struct Proof {
+    payload: ProofPayload,
+    signatures: Signatures,
+}
+
+impl Proof {
+    /// Constructor for `CommitTimeout` view change suggestion.
+    pub fn commit_timeout(
+        voting_block_hash: Hash,
+        previous_proof: Hash,
+        latest_block: Hash,
+    ) -> Self {
+        Self {
+            signatures: Signatures::default(),
+            payload: ProofPayload {
+                reason: Reason::CommitTimeout(CommitTimeout { voting_block_hash }),
+                previous_proof,
+                latest_block,
+            },
+        }
+    }
+
+    /// Constructor for `BlockCreationTimeout` view change suggestion.
+    pub fn block_creation_timeout(
+        transaction_receipt: TransactionReceipt,
+        previous_proof: Hash,
+        latest_block: Hash,
+    ) -> Self {
+        Self {
+            signatures: Signatures::default(),
+            payload: ProofPayload {
+                reason: Reason::BlockCreationTimeout(BlockCreationTimeout {
+                    transaction_receipt,
+                }),
+                previous_proof,
+                latest_block,
+            },
+        }
+    }
+
+    /// Constructor for `NoTransactionReceiptReceived` view change suggestion.
+    pub fn no_transaction_receipt_received(
+        transaction_hash: Hash,
+        previous_proof: Hash,
+        latest_block: Hash,
+    ) -> Self {
+        Self {
+            signatures: Signatures::default(),
+            payload: ProofPayload {
+                reason: Reason::NoTransactionReceiptReceived(NoTransactionReceiptReceived {
+                    transaction_hash,
+                }),
+                previous_proof,
+                latest_block,
+            },
+        }
+    }
+
+    /// Hash of the proof `payload`.
+    pub fn hash(&self) -> Hash {
+        let bytes: Vec<u8> = self.payload.clone().into();
+        Hash::new(&bytes)
+    }
```

Comment:
Payload is just 2 hashes and 1 structure. Can we do hashing wo allocation here?

Reply:
I am not sure, what do you mean? Can you provide a sample of code?

In general we need the whole payload in bytes to take its hash as the whole payload information is important to the consensus.

Reply:
I mean generally speaking rust default `Hash` trait just take `&self` and hashes object itself.
https://doc.rust-lang.org/std/hash/trait.Hash.html#tymethod.hash

So it doesn't require allocations in first place and generally it is implemented as derive macro which internally just hashes all fields by reading them without allocations

Reply:
Yes, rust default hashing is very different as they do not need consistency between different languages :) But we need it, therefore we use SCALE for byte serialization and then Ursa for hashing.

Reply:
Also we have a requirement that we use Ursa for all the cryptography.

Reply:
This is not about consistency, this point is about performance. We can do language independent hashing wo allocations. I mean this is just a thought, but hashing itself is costy and allocation and copying it seems kinda bad.